Fix Lwt_io.of_bytes

NathanReb opened the pull request:
Fixes #636

My understanding is that it's `offset` that's improperly set in `Lwt_io.of_bytes`. Since `Lwt_bytes` based channels just use the `Lwt_bytes` straight as their internal buffer and never actually perform any io, `offset` is never updated. That means we can consider that everything has been already read in case of an input channel since it's already in the internal buffer. In the output channel case, writing to the internal buffer is sufficient and we never need to flush meaning offset should be set to the end of the buffer here as well.

Let me know if that seems right to you!

I'd like to add some tests for this. They should be in `ls test/unix/test_lwt_io.ml` right ?
I was thinking about testing the initial position after creating a channel from bytes. Let me know if you have any brighter ideas for good tests here!

aantron commented:
Great, taking a look!

Yes, `test/unix/test_lwt_io.ml` is a good place for the tests. I would test the initial position, the position after some I/O is done, and the position after `set_position`, on both `Lwt_bytes` and ordinary channels. Feel free to write only a subset of those tests :)

aantron commented:
The reasoning seems correct for `Input`, but I think it will give the wrong result for `Output` (I guess it's important to vary `Input`/`Output` in the test cases as well).

`offset` seems to be the number of bytes either actually read into the buffer, or actually written *from* the buffer to the outside world, not *to* the buffer from the user (when `set_position` is not being used). That means that when an `Lwt_bytes` channel is first created, with this change, the position that will be reported will be the end of the buffer, but I think it should be zero instead.

Our writing-from-the-buffer operation is just a no-op in the `Lwt_bytes` case.

aantron commented:
Please disregard the noisy macOS build rows :) I need to adjust timeouts in some of the tests.

NathanReb commented:
Yes I just realized by writing the tests and reading the code further.
I did add a bunch of tests (which were definitely useful since I found out about the new bug I introduced), both for output and input but just for `Lwt_bytes` since the only function I modified was `of_bytes`.

I'll gladly add more tests for regular channels in a separate PR if find some time to do so!

NathanReb commented:
Is there anything blocking here? Can I do something to help get it merged?

aantron commented:
Nothing inherent to the PR :) I'm just continuously busy then traveling, but I should be able to review this fully and very likely merge later today, or early tomorrow.

aantron commented:
Thanks!
